sv.c

```diff
@@ -4771,6 +4776,272 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
+                    }
+                    CowREFCNT(ssv)++;
+                    SvPV_set(dsv, SvPVX_mutable(ssv));
+                    sv_buf_to_ro(ssv);
+                } else
+#endif
+                {
+                /* SvIsCOW_shared_hash */
+                DEBUG_C(PerlIO_printf(Perl_debug_log,
+                                      "Copy on write: Sharing hash\n"));
+                SvPV_set(dsv,
+                         HEK_KEY(share_hek_hek(SvSHARED_HEK_FROM_PV(SvPVX_const(ssv)))));
+                }
+                SvLEN_set(dsv, len);
+                SvCUR_set(dsv, cur);
+                SvIsCOW_on(dsv);
+            } else {
+                /* Failed the swipe test, and we cannot do copy-on-write either.
+                   Have to copy the string.  */
+                sv_grow_fresh(dsv, cur + 1);
+
+                Move(SvPVX_const(ssv),SvPVX(dsv),cur,char);
+                SvCUR_set(dsv, cur);
+                *SvEND(dsv) = '\0';
+            }
+        SvFLAGS(dsv) |= sflags & SVf_UTF8;
+}
+
+/* A helper for sv_mortalcopy_flags and newSVsv_flags, more efficient
+ * for their use cases compared to Perl_sv_setsv_flags. */
+PERL_STATIC_INLINE SV *
+S_sv_freshcopy_flags(pTHX_ SV* ssv, const I32 flags)
+{
+    /* Note: If a PVIV/PVNV/PVMG is only IOK, NOK, ROK, it is _mostly_
+     * possible to create just a headless SV to store that value.
+     * Some parts of core (Perl_amagic_call in gv.c specifically) do
+     * assume - and possibly CPAN might - that SvTYPE(dsv) == SvTYPE(ssv)
+     * though, which is why the code below does not try that type
+     * simplification. Perhaps this might be worth revisiting in the future.
+     * -- April 2025. */
+    assert(ssv);
+    svtype stype = SvTYPE(ssv);
+
+#if NVSIZE <= IVSIZE
+    ASSUME(stype > SVt_NV);
+#else
+    ASSUME(stype > SVt_IV);
+#endif
+    ASSUME(stype <= SVt_PVMG);
+    SV *dsv = newSV_type(stype);
+    U32 sflags   = SvFLAGS(ssv);
+
+    ASSUME(!(flags & SV_GMAGIC)); /* Caller should have sorted this. */
+    /* Note: Perl_sv_setsv_flags essentially ignores magic, except for
+     * taint and vstring magic. */
+
+    /* The switch cases are in the order that gcov of make/make test suggests. */
+    switch(sflags & (SVp_IOK|SVp_NOK|SVf_ROK|SVp_POK|SVf_FAKE|SVppv_STATIC)) {
+        case SVp_IOK:
+            SvIV_set(dsv, SvIVX(ssv));
+            SvFLAGS(dsv) |= sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV);
+            break;
+        case SVp_POK:
+            if (UNLIKELY(SvTYPE(ssv) == SVt_INVLIST)) {
+               invlist_clone(ssv, dsv);
+               return dsv;
+            }
+            S_sv_freshcopy_POK(aTHX_ dsv, ssv, flags);
+        {
+            const char *vstr_pv;
+            STRLEN vstr_len;
+            if ((vstr_pv = SvVSTRING(ssv, vstr_len))) {
+                sv_magic(dsv, NULL, PERL_MAGIC_vstring, vstr_pv, vstr_len);
+                SvRMAGICAL_on(dsv);
+            }
+        }
+            break;
```

Comment:
I see same-looking code on `case:` above and `case:` below this comment, maybe remove the` break;` and fall through?

Reply:
I am going to look into pulling the SvVSTRING code out to the caller, which would help.

It doesn't currently fall through `SVp_POK` into/from `SVp_POK|SVp_IOK` because:
* `SVp_POK` is the much more common case, so it would be a shame to fall through and  encounter integer-related branches that we know are irrelevant.
* `SVp_POK|SVp_IOK` can't easily fall through without also adding an integer-checking branch in `SVp_POK`'s case.

Reply:
nvm, I see your imputing 3^3 or 3*3 permutation of PINOK public flg and private flg as an optimization